### ISSUE TITLE
Connection reuse returning no error, but empty results

### ISSUE THREAD

**Issue by d3mon187** [go-redis/redis#1830]
I'm not very familiar with Go, but rewrote my api to use go-redis, so please excuse me if this is a simple mistake.

On my first attempt, I put the NewClient variable in my handler function outside of main, which works well aside from being super inefficient and using thousands of file descriptors.
Realizing my mistake, I moved it to the main function with a global var.  Now it works with reusing connections, but randomly it will start returning empty results with no errors!?   Which of course is horrible, because that registers as a valid result. I can force it to throw errors, and have gotten EOF errors in some cases with timeouts, but this has no error.   It happens whether I use Sentinel or my keepalived ip.  The only reliable method is to connect everytime handler function is run.

Here is a basic outline of the program.  Same

```
import (
	redis "github.com/go-redis/redis/v8"
)
var rdb *redis.Client
var ctx = context.Background()
func main() {
	router := http.NewServeMux()
	router.HandleFunc("/api", ApiHandler)
	srv := &http.Server{
		Addr:    ":8080",
		ReadTimeout:  5 * time.Second,
		WriteTimeout: 10 * time.Second,
	}
	rdb = redis.NewClient(&redis.Options{
			Addr: "192.168.1.2:6379",
			Password: "password",
			DB:       0,  // use default DB
			//DialTimeout:        10 * time.Second,
			//ReadTimeout:        30 * time.Second,
			//WriteTimeout:       30 * time.Second,
			PoolSize: 1000,
			//PoolTimeout:        30 * time.Second,
			//IdleTimeout:        time.Second * 25,
			IdleCheckFrequency: time.Second * 5,
			MaxRetries:         3,
	})
	log.Fatal(srv.ListenAndServe())
}
func ApiHandler(w http.ResponseWriter, r *http.Request) {
	key, ok := r.URL.Query()["key"]
	api_key := key[0]
	//works if I move rdb newclient here
	key_info, err := rdb.HGetAll(ctx, "key:" + api_key).Result()
	if err != nil {
		log.Println("key info error: ", err)
	} else  len(key_info) == 0 {
		log.Println("key info does not exist: ", err)
		log.Println("Missing apiKey is: " + api_key)
	} else {
		log.Println("Good apiKey is: " + api_key)
	}
	defer r.Body.Close()
	return
}
```

After many connections, the same key that registers as Good will randomly come up as Missing.  Sometimes once, sometimes many times, before stating Good again.   If I move the NewClient to my comment in ApiHandler, it is always Good.
Thanks!

**Comment by monkey92t** [comment]
It may be that the interval between multiple tests is too long (http), which causes the active connections in the connection pool to be closed by the server. We have recently added a new function to check whether the connection is healthy, and a new version will be released later. You can try it at that time

**Comment by d3mon187** [comment]
Thanks for the reply!  I think what happens is that sometimes multiple connections are started in the pool, and then at slower periods they are closed when only one is needed.

Either way, shouldn't it be throwing an error though?  Otherwise I can't really trust an empty or 0 result and would need to ping the server to make sure, correct?

Appreciate all the work!

**Comment by monkey92t** [comment]
I can't replicate your environment, can you make an example that triggers an empty result?  (err=nil)

**Comment by monkey92t** [comment]
I executed the above example, and I did not find any empty results.

**Comment by d3mon187** [comment]
Good news, I'm dumb.  

I knew I screwed up when my Ping test came back good, but results were still empty.  Then I realized my synchronous loving brain didn't grasp that in my full program, if I make a db change with rdb.Do(Select ,1), that the change would effect other requests running at the same time.  Doh!  Now I've made separate NewClients for each db connection, and all seems to be working well.  Not sure if there's a better way to do it, but having a couple clients isn't a problem seeing as I had 1000s the other way.

Thanks for the help!
